File: Matchismo OBJC/Model/PlayingCard.m
//
//  PlayhingCard.m
//  Matchismo OBJC
//
//  Created by Пермяков Андрей on 6/21/19.
//  Copyright © 2019 Пермяков Андрей. All rights reserved.
//

#import "PlayingCard.h"

@implementation PlayingCard


//some constants that we can change if we come up with a better counting algorithm
static const int MATCHED_2_CARDS_RANKS = 8;
static const int MATCHED_2_CARDS_RANKS_IN_3_CARDS_MODE = 6;
static const int MATCHED_3_CARDS_RANKS = 18;
static const int MATCHED_2_CARDS_SUITS = 2;
static const int MATCHED_2_CARDS_SUITS_IN_3_CARDS_MODE = 1;
static const int MATCHED_3_CARDS_SUITS = 8;

- (int)match:(NSArray *)otherCards {
    int score = 0;
    if ([otherCards count] == 1) {
        PlayingCard *otherCard = [otherCards firstObject];
        if (otherCard.rank == self.rank) {
            score = MATCHED_2_CARDS_RANKS;
        } else if ([otherCard.suit isEqualToString:self.suit]) {
            score = MATCHED_2_CARDS_SUITS;
        }
    } else if ([otherCards count] == 2) {
        PlayingCard *firstOtherCard = [otherCards firstObject];
        PlayingCard *secondOtherCard = [otherCards lastObject];
        int suitScore = 0;
        int rankScore = 0;
        if ([self.suit isEqualToString:firstOtherCard.suit] || [self.suit isEqualToString:secondOtherCard.suit]) {
            suitScore += MATCHED_2_CARDS_SUITS_IN_3_CARDS_MODE;
        }
        if ([firstOtherCard.suit isEqualToString:secondOtherCard.suit]) {
            suitScore = suitScore ? MATCHED_3_CARDS_SUITS : MATCHED_2_CARDS_SUITS_IN_3_CARDS_MODE;
        }
        
        if (self.rank == firstOtherCard.rank || self.rank == secondOtherCard.rank) {
            rankScore += MATCHED_2_CARDS_RANKS_IN_3_CARDS_MODE;
        }
        if (firstOtherCard.rank == secondOtherCard.rank) {
            rankScore = rankScore ? MATCHED_3_CARDS_RANKS : MATCHED_2_CARDS_RANKS_IN_3_CARDS_MODE;
        }
        score = suitScore + rankScore; // if no match, both are 0
    }
    return score;
}

- (NSString *)contents {
    NSArray *rankStrings = [PlayingCard rankStrings];
    return [rankStrings[self.rank] stringByAppendingString:self.suit];
}

@synthesize suit = _suit;

+ (NSArray *)validSuits {
    return @[@"♥", @"♠", @"♦", @"♣"];
}

- (void)setSuit:(NSString *)suit {
    if ([[PlayingCard validSuits] containsObject:suit]) {
        _suit = suit;
    }
}

- (NSString *)suit {
    return _suit ? _suit : @"?";
}

+ (NSArray *)rankStrings {
    return @[@"?", @"A", @"2", @"3", @"4", @"5", @"6", @"7", @"8", @"9", @"10", @"J", @"Q", @"K"];
}

+ (NSUInteger)maxRank { return [[self rankStrings] count] - 1; }

- (void)setRank:(NSUInteger)rank {
    if (rank <= [PlayingCard maxRank]) {
        _rank = rank;
    }
}

@end

File: Matchismo Swift/Model/PlayingCard.swift
//  Converted to Swift 5 by Swiftify v5.0.37171 - https://objectivec2swift.com/
//
//  PlayhingCard.swift
//  Matchismo
//
//  Created by Пермяков Андрей on 6/21/19.
//  Copyright © 2019 Пермяков Андрей. All rights reserved.
//

class PlayingCard: Card {
  private var _suit: String?
  var suit: String {
    get {
      if let _suit = _suit, _suit.count > 0 {
        return _suit
      } else {
        return "?"
      }
    }
    set(suit) {
      if PlayingCard.validSuits().contains(suit) {
        _suit = suit
      }
    }
  }
  
  private var _rank: Int?
  var rank: Int {
    get {
      return _rank ?? 0
    }
    set(rank) {
      if rank <= PlayingCard.maxRank() {
        _rank = rank
      }
    }
  }
  
  override var contents: String {
    get {
      let rankStrings = PlayingCard.rankStrings()
      return rankStrings[rank] + suit
    }
    set {}
  }
  
  class func validSuits() -> [String] {
    return ["♥", "♠", "♦", "♣"]
  }
  
  class func maxRank() -> Int {
    return self.rankStrings().count - 1
  }
  
  private class func rankStrings() -> [String] {
    return ["?","A","2","3","4","5","6","7","8","9","10","J","Q","K"]
  }
  
  // Returns 0 if no match, positive Int otherwise
  override func match(_ otherCards: [Card]) -> Int {
    guard let otherCards = otherCards as? [PlayingCard] else {
      return 0
    }
    return getScore(forRank: rank, comparingToCards: otherCards) +
      getScore(forSuit: suit, comparingToCards: otherCards)
  }
  
  private func getScore(forSuit suit: String, comparingToCards cards: [PlayingCard]) -> Int {
    var score = 0
    switch cards.count {
    case 1:
      if cards.first?.suit == suit {
        score = CardMatchingBounses.twoSuits
      }
    case 2:
      guard let firstOtherCard = cards.first,
        let secondOtherCard = cards.last else { break }
      if suit == firstOtherCard.suit || suit == secondOtherCard.suit {
        score += CardMatchingBounses.twoSuitsInThreeCardMode
      }
      if firstOtherCard.suit == secondOtherCard.suit {
        score = score != 0 ? CardMatchingBounses.threeSuits :
          CardMatchingBounses.twoSuitsInThreeCardMode
      }
    default:
      break
    }
    return score
  }
  
  private func getScore(forRank rank: Int, comparingToCards cards: [PlayingCard]) -> Int {
    var score = 0
    switch cards.count {
    case 1:
      if cards.first?.rank == rank {
        score = CardMatchingBounses.twoRanks
      }
    case 2:
      guard let firstOtherCard = cards.first,
        let secondOtherCard = cards.last else { break }
      if rank == firstOtherCard.rank || rank == secondOtherCard.rank {
        score += CardMatchingBounses.twoRanksInThreeCardMode
      }
      if firstOtherCard.rank == secondOtherCard.rank {
        score = score != 0 ? CardMatchingBounses.threeRanks :
          CardMatchingBounses.twoRanksInThreeCardMode
      }
    default:
      break
    }
    return score
  }
}
